### DISCUSSION TITLE
Tsserver and denolsp conflict

### DISCUSSION THREAD

**Discussion by Lite5h4dow** [williamboman/nvim-lsp-installer#597] [Q&A] [answered]
Hi. ive been trying to set up lsp and cmp for 2 days and one issue i keep coming back to is tsserver and denolsp conflicting on typescript projects ive looked at configs from a bunch of places ill list them here so you know where im coming from and what ive tried.

 - https://github.com/LunarVim/Neovim-from-scratch/blob/06-LSP/lua/user/lsp/lsp-installer.lua
 - https://github.com/neovim/nvim-lspconfig
 - https://github.com/VonHeikemen/lsp-zero.nvim
 - https://deno.land/manual@v1.20.5/getting_started/setup_your_environment#neovim-06-and-nvim-lspconfig

i cant seem to get any of these to work on their own or in concert and its killing my mind. please help. here are the important files for the configs.

`lua/lsp/init.lua`
```lua
--lsp zero
local lsp=require('lsp-zero')
lsp.preset("recommended")
lsp.nvim_workspace()
lsp.setup()

-- lsp install
local lsp_installer = require("nvim-lsp-installer")

lsp_installer.settings({
  install_root_dir = string.gsub((vim.fn.system("echo $HOME") .. "/.config/nvim/.lsp_servers"), "\n", "")
})

lsp_installer.on_server_ready(function(server)
  local opts={
    on_attach = require("lsp.handlers").on_attach,
    capabilities = require("lsp.handlers").capabilities
  }

  local opts_exist, server_opts = pcall(require, "lsp.settings.".. server.name)
  if opts_exist then
    opts = vim.tbl_deep_extend("force", server_opts, opts)
  end
  server:setup(opts)
end
)
```

`lua/lsp/handlers.lua`
```lua
local cmp_nvim_lsp = require("cmp_nvim_lsp")

local capabilities = vim.lsp.protocol.make_client_capabilities()

return {
  on_attach = function(client, bufnr)
    if client.name == "tsserver" then
      client.resolved_capabilities.document_formatting = false
    end
  end,
  capabilities = cmp_nvim_lsp.update_capabilities(capabilities),
}
```

`lua/lsp/settings/tsserver.lua`
```lua
local util = require "lspconfig.util"
local bin_name = 'typescript-language-server'
local cmd = { bin_name, '--stdio' }

return{
  default_config ={
    init_options={
      hostInfo = 'neovim'
    },
    cmd = cmd,
    filetypes = {
      "javascript",
      "javascriptreact",
      "javascript.jsx",
      "typescript",
      "typescriptreact",
      "typescript.tsx",
    },
    root_dir = function () util.root_pattern("package.json") end
  }
}
```

for full configs in context to the rest of neovim configs, please look here: https://github.com/Lite5h4dow/.dotfiles/tree/master/.config/nvim

⚠️i use yadm as my manager, please ignore the `##default/t` extensions they are exclusivly for yadm and generate / symlink to normal files, they dont change anything.

**Comment by williamboman** [comment]
Hello! I'm afraid I'm not following what exactly the problem is?

**Reply by Lite5h4dow** [reply]
the problem is the `default_config` table wrapping the actual config. its in lspconfig's examples, but its wrong and it drove me insane for 4 days. kill me.

**Comment by Lite5h4dow** [comment, marked as answer]
DONT WRAP YOUR CONFIGS IN `default_config`, LSPCONFIG IS LYING WITH THE EXAMPLES.

**Reply by williamboman** [reply]
Ah I see, that's annoying haha :D. In which examples did you see things being wrapped with `default_config`?

**Reply by Lite5h4dow** [reply]
this entire god damn folder
https://github.com/neovim/nvim-lspconfig/tree/master/lua/lspconfig/server_configurations

**Reply by williamboman** [reply]
Ah those aren't really examples, those are the source files. You can find the docs [here](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md) instead, for the future

**Reply by Lite5h4dow** [reply]
Yeah, the thing is, I didn't know what was supposed to go where and I was very confused. Hence why I found that folder and used it as an example.

**Comment by Decodetalkers** [comment]
```lua
lsp_installer.on_server_ready(function(server)
    elseif server.name == "denols" then
        --local settings = require("deno-tool").read()
        local opts = {
            on_attach = on_attach,
            root_dir = nvim_lsp.util.root_pattern("deno.json"),
            init_options = { --settings,
                lint = true,
            },
        }
        server:setup(opts)
    elseif server.name == "tsserver" then
        local opts = {
            on_attach = on_attach,
            root_dir = nvim_lsp.util.root_pattern("package.json"),
            init_options = {
                lint = true,
            },
        }
        server:setup(opts)
    else
        local opts = {
            on_attach = on_attach,
        }

        -- (optional) Customize the options passed to the server
        -- if server.name == "tsserver" then
        --     opts.root_dir = function() ... end
        -- end

        -- This setup() function is exactly the same as lspconfig's setup function.
        -- Refer to https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md
        server:setup(opts)
    end
end)
```
You can set Like this , just  config it like nvim-lspconfig wiki said, change the rootpattern

**Reply by Lite5h4dow** [reply]
The entire point is to do it programmatically without editing the main config every time.
